THREESCALE-10697: Add debug mode for the worker and add RubyMine configurations for both worker and listener (#363)
3scale/apisonator: DEVELOPMENT.md

be found in the manifest:

> `$ bundle exec 3scale_backend manifest`

## Running with RubyMine

Using RubyMine scripts can help you run both Worker and Listener.

For that, you'll need to create the following configurations: 

### Worker

> | Name                  | `worker` |

Reviewer: You would need to add a header for the table to make it look good. I tried this too at first (because there is no real need for the header), but it just doesn't work :grimacing: 

![image](https://github.com/3scale/apisonator/assets/1270328/4fd3cce5-445b-4417-bb81-5d28fd317d01)

| Ruby script           | `/path/to/project/apisonator/bin/3scale_backend_worker` |
| Script arguments      | `--debug` |
| Working directory     | `/path/to/project/apisonator` |
| Environment variables | `RACK_ENV=development` |

In case you want to run the worker in async mode you can add `CONFIG_REDIS_ASYNC=true` to the environment variables here or directly add it to your `.env` file.

### Puma Listener (sync mode)

| Name                  | `listener (puma)` |
| Ruby script           | `/path/to/project/apisonator/bin/3scale_backend` |
| Script arguments      | `-s puma -X start -p 3001` |
| Working directory     | `/path/to/project/apisonator` |

> | Environment variables | `RACK_ENV=development;LISTENER_WORKERS=1` |

Reviewer 2: ```suggestion
| Environment variables | `RACK_ENV=development` |
```

### Falcon Listener (async mode)

| Name                  | `listener (falcon)` |
| Ruby script           | `/path/to/project/apisonator/bin/3scale_backend` |
| Script arguments      | `-s falcon start -p 3001` |
| Working directory     | `/path/to/project/apisonator` |

> | Environment variables | `LISTENER_WORKERS=1;RACK_ENV=development;CONFIG_REDIS_ASYNC=true` |

Reviewer 2: ```suggestion
| Environment variables | `RACK_ENV=development;CONFIG_REDIS_ASYNC=true` |
```

be found in the manifest:

> `$ bundle exec 3scale_backend manifest`

> 
> ## Running with RubyMine
> 
> Using RubyMine scripts can help you run both Worker and Listener.

Reviewer 2: I don't like mentioning RubyMine in particular because it's something we use but the docs are for the public and I don't think much people use RubyMine since it's paid and not free software.

I suggest this rewording:

```suggestion

## Running from your IDE

Some IDEs like VSCode or RubyMine allow to add custom launchers for scripts. This is convenient to launch both Worker and Listener.
```

Author: Sounds good, applied!

be found in the manifest:

> `$ bundle exec 3scale_backend manifest`

## Running with RubyMine

Using RubyMine scripts can help you run both Worker and Listener.

For that, you'll need to create the following configurations: 

> ### Worker

Reviewer: For the worker we can also mention that the sync/async mode is configured by the `CONFIG_REDIS_ASYNC` in `.env`.

First you'll have to set a name. We recommend using `worker`.

Then you'll have to add the following config:

```

> Ruby script: /Users/lvillenj/Projects/apisonator/bin/3scale_backend_worker

Reviewer: I'd replace your path with some example path, and a comment that this will need to be replaced (in the "parent" section, because it will apply to all configs).

Author: Totally correct, I don't know why I skipped this, thank you 🙏

Script arguments: --debug
Working directory: /Users/lvillenj/Projects/apisonator
Environment variables: RACK_ENV=development
```

### Puma Listener

First you'll have to set a name. We recommend using `listener (puma)`.

Then you'll have to add the following config:

```
Ruby script: /Users/lvillenj/Projects/apisonator/bin/3scale_backend
Script arguments: -s puma -X "-w 0" start -p 3001
Working directory: /Users/lvillenj/Projects/apisonator

> Environment variables: RACK_ENV=development;LISTENER_WORKERS=1

Reviewer: Does `LISTENER_WORKERS` do anything? I think it doesn't change the behavior for me, `"-w 0"` already sets puma server in "single mode", so there are actually no additional workers created. I think I'd remove it.

Reviewer 2: IMO it does nothing

```

> ### Falcon Listener

Reviewer: Maybe a comment that falcon listener is for async mode could be useful.

> ### Falcon Listener

Reviewer 3: Nice! What is this used for though? Maybe a line of explanation why and when we want to run each one of those.

Author: These configs are the ones you would use with RubyMine configurations and they just intend to make life easier for developers, for more details users can go to `README.md` I'd say, WDYT?

Reviewer 3: Ok, maybe actually link from README to DEVELOPMENT about these. The README instructions about `make` seem to be broken though. I hope the `docker` instructions aren't.

Reviewer 2: @akostadinov The listener runs Puma on sync mode and Falcon on async mode.
